filestore/util.go

```diff
@@ -0,0 +1,175 @@
+package filestore
+
+import (
+	"fmt"
+	//"io"
+	//"os"
+	//"path/filepath"
+
+	//"github.com/ipfs/go-ipfs/blocks"
+	"github.com/ipfs/go-ipfs/blocks/blockstore"
+	pb "github.com/ipfs/go-ipfs/filestore/pb"
+	dshelp "github.com/ipfs/go-ipfs/thirdparty/ds-help"
+	//posinfo "github.com/ipfs/go-ipfs/thirdparty/posinfo"
+
+	ds "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore"
+	//dsns "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/namespace"
+	dsq "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/query"
+	proto "gx/ipfs/QmT6n4mspWYEya864BhCUJEgyxiRfmiSY9ruQwTUNpRKaM/protobuf/proto"
+	cid "gx/ipfs/QmcTcsTvfaeEBRFo1TkFgT8sRmgi1n1LTZpecfVP8fzpGD/go-cid"
+)
+
+type Status int32
+
+const (
+	StatusOk           Status = 0
+	StatusFileError    Status = 10 // Backing File Error
+	StatusFileNotFound Status = 11 // Backing File Not Found
+	StatusFileChanged  Status = 12 // Contents of the file changed
+	StatusOtherError   Status = 20 // Internal Error, likely corrupt entry
+	StatusKeyNotFound  Status = 30
+)
+
+func (s Status) String() string {
+	switch s {
+	case StatusOk:
+		return "ok"
+	case StatusFileError:
+		return "error"
+	case StatusFileNotFound:
+		return "no-file"
+	case StatusFileChanged:
+		return "changed"
+	case StatusOtherError:
+		return "ERROR"
+	case StatusKeyNotFound:
+		return "missing"
+	default:
+		return "???"
+	}
+}
+
+func (s Status) Format() string {
+	return fmt.Sprintf("%-7s", s.String())
+}
+
+type ListRes struct {
+	Status   Status
+	ErrorMsg string
+	Key      *cid.Cid
+	FilePath string
+	Offset   uint64
+	Size     uint64
+}
+
+func (r *ListRes) FormatLong() string {
+	switch {
+	case r.Key == nil:
+		return "?????????????????????????????????????????????????"
```

Comment:
It might happen, if the key is corrupt, in which case there will also be an error.  The idea is to not abort and be able to move on to the next entry.